lib/KarmaServer.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * KarmaServer class that contains all functionality for managing a Karma server. This includes starting the server,
+ * stopping the server, and running tests on the server.
+ *
+ * @author Jimi van der Woning
+ */
+'use strict';
+
+var _ = require('lodash'),
+    path = require('path'),
+    Q = require('q'),
+    spawn = require('child_process').spawn,
+    KarmaServerStatus = require('./KarmaServerStatus');
+
+// The Karma test runner
+var runner;
+
+
+/**
+ * Constructor for a Karma server instance
+ *
+ * @param config {object} Karma configuration object that should be used
+ * @param port {number} The port on which the Karma server should run
+ * @constructor
+ */
+function KarmaServer(config, port) {
+    this._status = null;
+    this._serverProcess = null;
+    this._config = _.merge({ waitForServerTime: 10, waitForRunnerTime: 2 }, config, { port: port });
+
+    runner = require('karma').runner;
+}
+
+/**
+ * Set the status of the server instance to the given status.
+ *
+ * @param status {number} The new status the server instance should have
+ * @private
+ */
+KarmaServer.prototype._setStatus = function(status) {
+    this._status = status;
+};
+
+/**
+ * Start the Karma server instance, if it has not been started before
+ *
+ * @returns {*|promise} a promise that will resolve with the instance itself when the instance starts properly within
+ *                      config.waitForServerTime seconds, and reject otherwise
+ */
+KarmaServer.prototype.start = function() {
+    var self = this,
+        deferred = Q.defer(),
+        startTime = Date.now(),
+        serverTimeout;
+
+    // Only servers that have not yet been started can be started
+    if(self._status === null) {
+        console.log('Starting a Karma server on port ' + self._config.port + '...');
+        self._setStatus(KarmaServerStatus.INITIALIZING);
+
+        // Limit the time it can take for a server to start to config.waitForServerTime seconds
+        serverTimeout = setTimeout(function() {
+            self._setStatus(KarmaServerStatus.DEFUNCT);
+            deferred.reject(
+                'Could not connect to a Karma server on port ' + self._config.port + ' within ' +
+                self._config.waitForServerTime + ' seconds'
+            );
+        }, self._config.waitForServerTime * 1000);
+
+        // Start a new Karma server process
+        self._serverProcess = spawn(
+            'node',
+            [
+                path.join(__dirname, 'run-karma-in-background.js'),
+                JSON.stringify(self._config)
+            ],
+            { stdio: 'pipe' }
+        );
+
+        // Handle data the Karma server writes to stdout
+        self._serverProcess.stdout.on('data', function(data) {
+            var message = data.toString('utf-8');
+
+            // This is a hack: Karma exposes no method of determining when the server is started up. We dissect the logs
+            if(message.indexOf('Connected on socket') > -1) {
```

Comment:
Did you take into account that the karma config may contain more than one browsers? In that case Karma will spit out one of the above lines for each server. I think that we should only start testing once all registered browsers are up.
Compare that to my version of the KarmaServerManager:
https://github.com/jimivdw/grunt-mutation-testing/blob/kill_looping_karma_servers/lib/KarmaServerManager.js